libdrgn/program.c

```diff
@@ -872,30 +880,81 @@ drgn_thread_iterator_create(struct drgn_program *prog,
 			    struct drgn_thread_iterator **ret)
 {
 	struct drgn_error *err;
-	err = drgn_program_cache_threads(prog);
-	if (err)
-		return err;
-
-	*ret = malloc(sizeof(struct drgn_thread_iterator));
+	*ret = malloc(sizeof(**ret));
 	if (!*ret)
 		return &drgn_enomem;
-	(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
-	return NULL;
+	(*ret)->prog_is_live = prog->flags & DRGN_PROGRAM_IS_LIVE;
+	if ((*ret)->prog_is_live) {
+		if (!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL)) {
+			err = drgn_error_create(
+				DRGN_ERROR_INVALID_ARGUMENT,
+				"iterating over the threads of a live userspace process is not currently supported");
+			goto out;
+		}
+		struct drgn_object ns;
+		drgn_object_init(&ns, prog);
+		err = drgn_program_find_object(prog, "init_pid_ns", NULL,
+					       DRGN_FIND_OBJECT_VARIABLE, &ns);
+		if (err)
+			goto out;
+		err = drgn_object_address_of(&ns, &ns);
+		if (err)
+			goto out;
+		err = task_iter_init(prog, &ns, &(*ret)->task_iterator);
+		if (err)
+			goto out;
+	} else {
+		err = drgn_program_cache_threads(prog);
+		if (err)
+			goto out;
+		(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
+	}
+
+out:
+	if (err)
+		free(*ret);
+	return err;
 }
 
 LIBDRGN_PUBLIC void
 drgn_thread_iterator_destroy(struct drgn_thread_iterator *it)
 {
+	if (it->prog_is_live)
+		task_iter_deinit(&it->task_iterator);
 	free(it);
 }
 
 LIBDRGN_PUBLIC struct drgn_error *
 drgn_thread_iterator_next(struct drgn_thread_iterator *it,
 			  struct drgn_thread **ret)
 {
-	*ret = it->iterator.entry;
-	if (it->iterator.entry)
-		it->iterator = drgn_thread_set_next(it->iterator);
+	if (it->prog_is_live) {
+		task_iter_entry_t *entry;
+		struct drgn_error *err =
+			task_iter_next(&it->task_iterator, &entry);
+		if (err)
+			return err;
+		if (!entry) {
+			*ret = NULL;
+			return NULL;
+		}
+		it->entry.thread = *entry;
+		struct drgn_object tid;
+		drgn_object_init(&tid, drgn_object_program(entry));
+		err = drgn_object_member_dereference(&tid, entry, "pid");
+		if (err)
+			return err;
+		union drgn_value tid_value;
+		err = drgn_object_read_integer(&tid, &tid_value);
+		if (err)
+			return err;
+		it->entry.tid = tid_value.uvalue;
+		*ret = &it->entry;
```

Comment:
Since we're returning a pointer to a structure in the iterator, the returned thread is only valid until `_next` is called again, correct? We should document that.

Reply:
Actually, we probably don't want to overwrite the thread for every call. The caller might want to save one or more threads to do something with later. E.g.,

```python
interesting_threads = []
for thread in prog.threads():
    if thread_is_interesting(thread):
        interesting_threads.append(thread)
do_something(interesting_threads)
```

The easiest way to handle this would be to `malloc` the thread, and then have a `drgn_thread_destroy()` function that must be called for each thread. You could have a flag on the PRSTATUS-based threads so that you don't need to allocate them separately and `drgn_thread_destroy()` is a no-op.

Reply:
I see what you mean, but I think a cleaner way to accomplish the same thing would be to have the python `Thread` contain a `struct drgn_thread` instead of a `struct drgn_thread*`, this way we can avoid making a separate allocation (and simply dereferencing the `struct drgn_thread*` received from `drgn_thread_iterator_next` in order to copy it into the `Thread` is just fine since the internal `struct drgn_object` gets copied via `drgn_object_copy` from the return value of `task_iter_next`).

Reply:
I'm currently trying to understand why making a list of the returned values actually works just fine right now instead of segfaulting, because as you point out there should be an issue with this.

Reply:
For the time being I implemented it using the method I mentioned, but I've left it as a separate commit so that I can easily undo it if you'd rather me implement it by `malloc`'ing the threads as you originally suggested.

Reply:
Resolved via video call, I will reply to this thread when I have implemented the discussed changes.

Reply:
Alright, new design has been implemented. I've left it as a separate commit for the time being for ease of review.